core/src/main/java/org/infinispan/conflict/resolution/DefaultConflictResolutionManager.java

```diff
@@ -0,0 +1,177 @@
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Consumer;
+import java.util.function.Predicate;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+import org.infinispan.Cache;
+import org.infinispan.commands.CommandsFactory;
+import org.infinispan.commands.remote.ClusteredGetCommand;
+import org.infinispan.commons.CacheException;
+import org.infinispan.commons.util.EnumUtil;
+import org.infinispan.container.DataContainer;
+import org.infinispan.container.entries.InternalCacheEntry;
+import org.infinispan.container.entries.InternalCacheValue;
+import org.infinispan.context.impl.FlagBitSets;
+import org.infinispan.distribution.DistributionManager;
+import org.infinispan.factories.annotations.Inject;
+import org.infinispan.remoting.inboundhandler.DeliverOrder;
+import org.infinispan.remoting.responses.Response;
+import org.infinispan.remoting.responses.SuccessfulResponse;
+import org.infinispan.remoting.rpc.ResponseMode;
+import org.infinispan.remoting.rpc.RpcManager;
+import org.infinispan.remoting.rpc.RpcOptions;
+import org.infinispan.remoting.transport.Address;
+import org.infinispan.statetransfer.StateTransferManager;
+import org.infinispan.topology.RebalancingStatus;
+import org.infinispan.util.logging.Log;
+import org.infinispan.util.logging.LogFactory;
+
+/**
+ * @author Ryan Emerson
+ */
+public class DefaultConflictResolutionManager<K, V> implements ConflictResolutionManager<K, V> {
+
+   private static Log log = LogFactory.getLog(DefaultConflictResolutionManager.class);
+
+   private Cache cache;
+   private CommandsFactory commandsFactory;
+   private DataContainer<K, V> dataContainer;
+   private DistributionManager distributionManager;
+   private StateReceiver<K, V> stateReceiver;
+   private RpcManager rpcManager;
+   private StateTransferManager stateTransferManager;
+   private final AtomicBoolean streamInProgress = new AtomicBoolean();
+
+   @Inject
+   public void inject(Cache cache,
+                      CommandsFactory commandsFactory,
+                      DataContainer<K, V> dataContainer,
+                      DistributionManager distributionManager,
+                      StateReceiver<K, V> stateReceiver,
+                      RpcManager rpcManager,
+                      StateTransferManager stateTransferManager) {
+      this.cache = cache;
+      this.commandsFactory = commandsFactory;
+      this.dataContainer = dataContainer;
+      this.distributionManager = distributionManager;
+      this.stateReceiver = stateReceiver;
+      this.rpcManager = rpcManager;
+      this.stateTransferManager = stateTransferManager;
+   }
+
+   @Override
+   public Map<Address, InternalCacheValue<V>> getAllVersions(K key) {
+      if (stateTransferManager.isStateTransferInProgressForKey(key))
+         throw log.getConflictStateTransferInProgress(key);
+
+      List<Address> keyReplicaNodes = distributionManager.getWriteConsistentHash().locateOwners(key);
```

Comment:
FYI (doesn't need a change): if you decide to ban operation during ST, writeCH will be identical to readCH. Using writeCH has the advantage of being able to find all owners, including those we can't read from.